core/src/main/java/hivemall/mf/CofactorModel.java

```diff
@@ -0,0 +1,629 @@
+        }
+        return result;
+    }
+
+
+    protected static RealVector calculateRSD(Feature thisItem, List<Feature> trainableItems, int numFactors,
+                                    Map<String, Double> fixedBias, Map<String, Double> changingBias, Map<String, RealVector> weights) {
+
+        String i = thisItem.getFeature();
+        double b = getBias(i, fixedBias);
+
+        RealVector accumulator = new ArrayRealVector(numFactors);
+
+        // m_ij is named the same as in cofacto.py
+        for (Feature cooccurrence : trainableItems) {
+            String j = cooccurrence.getFeature();
+            double scale = cooccurrence.getValue() - b - getBias(j, changingBias);
+            RealVector g = getFactorVector(j, weights);
+            addInPlace(accumulator, g, scale);
+        }
+        return accumulator;
+    }
+
+    /**
+     * Calculate W' x W plus regularization matrix
+     */
+    protected static RealMatrix calculateWTWpR(Map<String, RealVector> W, int numFactors, float c0, RealMatrix idMatrix, float lambda) {
+        RealMatrix WTW = calculateWTW(W, numFactors, c0);
+        RealMatrix R = calculateR(idMatrix, lambda, numFactors);
+        return WTW.add(R);
+    }
+
+    protected static RealMatrix calculateR(RealMatrix idMatrix, float lambda, int numFactors) {
+        if (idMatrix == null) {
+            idMatrix = new Array2DRowRealMatrix(MatrixUtils.eye(numFactors));
+        }
+        return idMatrix.scalarMultiply(lambda);
+    }
+
+    protected static List<Feature> filterTrainableFeatures(Feature[] features, Map<String, RealVector> weights) {
+        List<Feature> trainableFeatures = new ArrayList<>();
+        String fName;
+        for (Feature f : features) {
+            fName = f.getFeature();
+            if (isTrainable(fName, weights)) {
+                trainableFeatures.add(f);
+            }
+        }
+        return trainableFeatures;
+    }
+
+    protected static RealVector solve(RealMatrix B, RealVector a) {
+        // b * x = a
+        // solves for x
+        SingularValueDecomposition svd = new SingularValueDecomposition(B);
+        return svd.getSolver().solve(a);
+
+    }
+
+    protected static RealMatrix calculateWTW(Map<String, RealVector> weights, int numFactors, float constant) {
+        RealMatrix WTW = new Array2DRowRealMatrix(numFactors, numFactors);
+        int i = 0, j = 0;
+        for (int f = 0; f < numFactors; f++) {
+            for (int ff = 0; ff < numFactors; ff++) {
+                double val = constant * dotFactorsAlongDims(weights, f, ff);
+                WTW.setEntry(f, ff, val);
+            }
+        }
+        return WTW;
+    }
+
+    protected static RealMatrix calculateWTWSubset(List<Feature> subset, Map<String, RealVector> weights, int numFactors, float constant) {
+        // equivalent to `B_u.T.dot((c1 - c0) * B_u)` in cofacto.py
+        RealMatrix delta = new Array2DRowRealMatrix(numFactors, numFactors);
+        int i = 0, j = 0;
```

Comment:
` int i = 0, j = 0;` never used. 

This could be found in IDE warning. Please resolve all IDE warnings.